pkcs7/benchmark-streaming-envelop.c

```diff
@@ -0,0 +1,397 @@
+    }
+
+    /* open the IO files to use */
+    if (ret == 0) {
+        io.in  = fopen(CONTENT_FILE_NAME, "rb");
+        io.out = fopen(ENCODED_FILE_NAME, "wb");
+        if (io.in == NULL || io.out == NULL) {
+            printf("Failed to open the IO files\n");
+            ret = -1;
+        }
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_SetStreamMode(pkcs7, 1, GetContentCB, StreamOutputCB,
+            (void*)&io);
+        if (ret != 0) {
+            printf("Failed to set stream mode\n");
+        }
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_EncodeEnvelopedData(pkcs7, NULL, 0);
+        if (ret <= 0) {
+            printf("Failed to encode enveloped data\n");
+        }
+    }
+
+    if (ret > 0) {
+        per = GetMBs(ret);
+        printf("%.2f MB/s", per);
+    }
+    printf(" : ret = %d\n", ret);
+
+    if (io.out != NULL) {
+        fseek(io.out, 0, SEEK_END);
+        printf("Created file [%s] with size of %ld bytes\n", ENCODED_FILE_NAME,
+            ftell(io.out));
+        fclose(io.out);
+    }
+
+    if (io.in != NULL) {
+        fclose(io.in);
+    }
+    wc_PKCS7_Free(pkcs7);
+
+    return 0;
+}
+
+
+static int DecryptCB(wc_PKCS7* pkcs7,
+    const byte* output, word32 outputSz, void* ctx) {
+     FILE* out = (FILE*)ctx;
+
+    if (out == NULL) {
+        return -1;
+    }
+
+    /* printf("Decoded in %d bytes\n", outputSz);
+     * for (word32 z = 0; z < outputSz; z++) printf("%02X", output[z]);
+     * printf("\n");
+    */
+    fwrite(output, 1, outputSz, out);
+
+    (void)pkcs7;
+    return 0;
+}
+
+static int DecodePKCS7Bundle(void)
+{
+    wc_PKCS7* pkcs7 = NULL;
+    double per;
+    int ret = 0;
+    FILE* f = NULL;
+    FILE* out = NULL;
+    double totalSz = 0;
+    byte testStreamBuffer[TEST_STREAM_CHUNK_SIZE];
+    int testStreamBufferSz = 0;
+
+    if (ret == 0) {
+        f = fopen(ENCODED_FILE_NAME, "rb");
+        if (f == NULL) {
+            printf("Unable to open encoded file\n");
+            ret = -1;
+        }
+        else {
+            fseek(f, 0, SEEK_END);
+        }
+    }
+
+    printf("\nDecoding bundle [%s], size of %ld bytes ... ",
+        ENCODED_FILE_NAME, ftell(f));
+    TimeLogStart();
+
+    pkcs7 = wc_PKCS7_New(NULL, 0);
+    if (pkcs7 == NULL) {
+        ret = MEMORY_E;
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_InitWithCert(pkcs7, (byte*)client_cert_der_2048,
+            sizeof_client_cert_der_2048);
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_SetKey(pkcs7, (byte*)client_key_der_2048,
+            sizeof_client_key_der_2048);
+    }
+
+    if (ret == 0) {
+        out = fopen(DECODED_FILE_NAME, "wb");
+        if (out == NULL) {
+            printf("Unable to open decrypted data out file\n");
+            ret = -1;
+        }
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_SetStreamMode(pkcs7, 1, NULL, DecryptCB, (void*)out);
+    }
+
+    if (ret == 0) {
+        rewind(f); /* start from the beginning of the file */
```

Comment:
Why are we using files?
If we are benchmarking just our code then we should be able to copy the data from another buffer unless the data is too large.

Reply:
OK, size of file is a parameter.